{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quickstart\n",
    "======"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is a small example on how to do Bayesian model selection."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "raw_mimetype": "text/restructuredtext"
   },
   "source": [
    "There are more examples in the `examples section <../examples.rst>`_ of the documentation, such as a `parameter inference example <parameter_inference.ipynb>`_ with a single model only.\n",
    "\n",
    "The notebook can be downloaded here:\n",
    ":download:`Quickstart <quickstart.ipynb>`.\n",
    "\n",
    "\n",
    "The following classes from the pyABC package are used\n",
    "for this example:\n",
    "\n",
    "* :class:`ABCSMC <pyabc.smc.ABCSMC>`\n",
    "* :class:`RV <pyabc.random_variables.RV>`\n",
    "* :class:`Distribution <pyabc.random_variables.Distribution>`\n",
    "* :class:`PercentileDistanceFunction <pyabc.distance_functions.PercentileDistanceFunction>`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step by step explanation\n",
    "----------------------------\n",
    "\n",
    "### Defining a model\n",
    "\n",
    "\n",
    "To do model selection, we first need some models. A model, in the simplest case,\n",
    "is just a callable which takes a single `dict` as input and returns a single `dict` as output. The keys of the input dictionary are the parameters of the model, the output\n",
    "keys denote the summary statistics.\n",
    "Here, the `dict` is passed as `parameters` and has the entry `x`, which denotes the mean of a Gaussian.\n",
    "It returns the observed summary statistics `y`, which is just the sampled value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "sys.path.append('../../')\n",
    "from model_comparison.utils import *\n",
    "from model_comparison.mdns import *\n",
    "from model_comparison.models import PoissonModel, NegativeBinomialModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.0\n"
     ]
    }
   ],
   "source": [
    "sample_size = 5\n",
    "ntest = 10\n",
    "\n",
    "k2 = 2.\n",
    "theta2 = 1.0\n",
    "\n",
    "k3 = 2.\n",
    "theta3 = 2. \n",
    "\n",
    "# then the scale of the Gamma prior for the Poisson is given by\n",
    "theta1 = 2.0\n",
    "k1 = (k2 * theta2 * k3 * theta3) / theta1\n",
    "print(k1)\n",
    "\n",
    "\n",
    "model_poisson = PoissonModel(sample_size=sample_size, n_workers=2)\n",
    "model_nb = NegativeBinomialModel(sample_size=sample_size, n_workers=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "import tempfile\n",
    "\n",
    "import scipy.stats as st\n",
    "\n",
    "from pyabc import (ABCSMC, RV, Distribution,\n",
    "                   PercentileDistanceFunction)\n",
    "\n",
    "# Define a gaussian model\n",
    "sigma = 1.\n",
    "\n",
    "\n",
    "def Poisson_model(parameters):\n",
    "    x = model_poisson.gen([parameters.lam])\n",
    "    return {'y': np.array([x.mean(), x.var()])}\n",
    "\n",
    "def NB_model(parameters): \n",
    "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
    "    return {'y': np.array([x.mean(), x.var()])}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For model selection we usually have more than one model.\n",
    "These are assembled in a list. We\n",
    "require a Bayesian prior over the models.\n",
    "The default is to have a uniform prior over the model classes.\n",
    "This concludes the model definition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We define two models, but they are identical so far\n",
    "models = [Poisson_model, NB_model]\n",
    "\n",
    "\n",
    "# Define priors \n",
    "prior1 = Distribution.from_dictionary_of_dictionaries(dict(lam={'type': 'gamma', 'kwargs': {'a':k1, 'scale': theta1}}))\n",
    "\n",
    "prior2 = Distribution.from_dictionary_of_dictionaries(dict(k={'type': 'gamma', 'kwargs': {'a':k2, 'scale': theta2}}, \n",
    "                                                     theta={'type': 'gamma', 'kwargs': {'a':k3, 'scale': theta3}}))\n",
    "\n",
    "parameter_priors = [prior1, prior2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configuring the ABCSMC run\n",
    "\n",
    "Having the models defined, we can plug together the `ABCSMC` class.\n",
    "We need a distance function,\n",
    "to measure the distance of obtained samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We plug all the ABC options together\n",
    "abc = ABCSMC(\n",
    "    models, parameter_priors,\n",
    "    PercentileDistanceFunction(measures_to_use=[\"y\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting the observed data\n",
    "\n",
    "Actually measured data can now be passed to the ABCSMC.\n",
    "This is set via the `new` method, indicating that we start\n",
    "a new run as opposed to resuming a stored run (see the \"resume stored run\" example).\n",
    "Moreover, we have to set the output database where the ABC-SMC run\n",
    "is logged."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:History:Start <ABCSMC(id=1, start_time=2018-12-13 10:29:27.729556, end_time=None)>\n",
      "Process Process-1:\n",
      "Process Process-2:\n",
      "Traceback (most recent call last):\n",
      "Process Process-3:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "Process Process-6:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "Process Process-8:\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "Process Process-7:\n",
      "Process Process-4:\n",
      "Process Process-5:\n",
      "Process Process-9:\n",
      "Process Process-10:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "AssertionError: daemonic processes are not allowed to have children\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Process Process-11:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Process Process-12:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "Process Process-13:\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "Process Process-14:\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "Process Process-15:\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Process Process-16:\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "Process Process-17:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "Process Process-18:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "Process Process-19:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "Process Process-20:\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Process Process-21:\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "Traceback (most recent call last):\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "Process Process-22:\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 15, in Poisson_model\n",
      "    x = model_poisson.gen([parameters.lam])\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "Process Process-23:\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "Process Process-24:\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\", line 24, in work\n",
      "    new_sim = simulate_one()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\", line 437, in simulate_one\n",
      "    t, theta, self.summary_statistics)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 115, in summary_statistics\n",
      "    raw_data = self.sample(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/model.py\", line 259, in sample\n",
      "    return self.sample_function(pars)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"<ipython-input-3-da275ed490b4>\", line 19, in NB_model\n",
      "    x = model_nb.gen([[parameters.k, parameters.theta]])\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "  File \"../../model_comparison/models.py\", line 32, in gen\n",
      "    p = Pool(processes=self.n_workers)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/context.py\", line 119, in Pool\n",
      "    context=self.get_context())\n",
      "AssertionError: daemonic processes are not allowed to have children\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 177, in __init__\n",
      "    self._repopulate_pool()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 238, in _repopulate_pool\n",
      "    self._wrap_exception)\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/pool.py\", line 257, in _repopulate_pool_static\n",
      "    w.start()\n",
      "  File \"/home/janfb/anaconda3/envs/sbi/lib/python3.7/multiprocessing/process.py\", line 110, in start\n",
      "    'daemonic processes are not allowed to have children'\n",
      "AssertionError: daemonic processes are not allowed to have children\n"
     ]
    },
    {
     "ename": "ProcessError",
     "evalue": "At least one worker is dead.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmpty\u001b[0m                                     Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicorebase.py\u001b[0m in \u001b[0;36mget_if_worker_healthy\u001b[0;34m(workers, queue)\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m             \u001b[0mitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mqueue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/multiprocessing/queues.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m    104\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_poll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 105\u001b[0;31m                         \u001b[0;32mraise\u001b[0m \u001b[0mEmpty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    106\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_poll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mEmpty\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mProcessError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-907fdbddb6ca>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m db_path = (\"sqlite:///\" +\n\u001b[1;32m      5\u001b[0m            os.path.join(tempfile.gettempdir(), \"test.db\"))\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mabc_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mabc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdb_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"y\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0my_observed\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\u001b[0m in \u001b[0;36mnew\u001b[0;34m(self, db, observed_sum_stat, gt_model, gt_par, meta_info)\u001b[0m\n\u001b[1;32m    352\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    353\u001b[0m         \u001b[0;31m# sample from prior to calibrate distance function and epsilon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 354\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_initialize_dist_and_eps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhistory\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_t\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    355\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    356\u001b[0m         \u001b[0;31m# return id generated in store_initial_data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\u001b[0m in \u001b[0;36m_initialize_dist_and_eps\u001b[0;34m(self, t)\u001b[0m\n\u001b[1;32m    378\u001b[0m             \u001b[0;31m# initialize distance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    379\u001b[0m             self.distance_function.initialize(t,\n\u001b[0;32m--> 380\u001b[0;31m                                               self._get_initial_samples(t)[1])\n\u001b[0m\u001b[1;32m    381\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    382\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meps\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequire_initialize\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\u001b[0m in \u001b[0;36m_get_initial_samples\u001b[0;34m(self, t)\u001b[0m\n\u001b[1;32m    416\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_initial_sum_stats\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msum_stats\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    417\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 418\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_initial_sum_stats\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sample_from_prior\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    419\u001b[0m                 self._initial_weights = [1.0 / len(self._initial_sum_stats)\n\u001b[1;32m    420\u001b[0m                                          for _ in self._initial_sum_stats]\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/smc.py\u001b[0m in \u001b[0;36m_sample_from_prior\u001b[0;34m(self, t)\u001b[0m\n\u001b[1;32m    447\u001b[0m         \u001b[0;31m# call sampler\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    448\u001b[0m         sample = self.sampler.sample_until_n_accepted(\n\u001b[0;32m--> 449\u001b[0;31m             self.population_strategy.nr_particles, simulate_one)\n\u001b[0m\u001b[1;32m    450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m         \u001b[0;31m# return all generated summary statistics\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicore_evaluation_parallel.py\u001b[0m in \u001b[0;36msample_until_n_accepted\u001b[0;34m(self, n, simulate_one)\u001b[0m\n\u001b[1;32m    109\u001b[0m         \u001b[0mn_done\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mn_done\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocesses\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 111\u001b[0;31m             \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_if_worker_healthy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocesses\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mqueue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    112\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mval\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mDONE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m                 \u001b[0mn_done\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/sbi/lib/python3.7/site-packages/pyabc/sampler/multicorebase.py\u001b[0m in \u001b[0;36mget_if_worker_healthy\u001b[0;34m(workers, queue)\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mEmpty\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mhealthy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mworkers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"At least one worker is dead.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m     \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The code should never reach here\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mProcessError\u001b[0m: At least one worker is dead."
     ]
    }
   ],
   "source": [
    "# y_observed is the important piece here: our actual observation.\n",
    "y_observed = [ 3.4 ,  5.04]\n",
    "# and we define where to store the results\n",
    "db_path = (\"sqlite:///\" +\n",
    "           os.path.join(tempfile.gettempdir(), \"test.db\"))\n",
    "abc_id = abc.new(db_path, {\"y\": y_observed})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `new` method returns an id, which is the id of the\n",
    "ABC-SMC run in the database.\n",
    "We're not usint this id for now.\n",
    "But it might be important when you load the stored data or want\n",
    "to continue an ABC-SMC run in the case of having more than one\n",
    "ABC-SMC run stored in a single database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"ABC-SMC run ID:\", abc_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running the ABC\n",
    "\n",
    "We run the `ABCSMC` specifying the epsilon value at which to terminate.\n",
    "The default epsilon strategy is the `pyabc.epsilon.MedianEpsilon`.\n",
    "Whatever is reached first, the epsilon or the maximum number allowed populations,\n",
    "terminates the ABC run. The method returns a `pyabc.storage.History` object, which\n",
    "can, for example, be queried for the posterior probabilities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We run the ABC until either criterion is met\n",
    "history = abc.run(minimum_epsilon=0.0001, max_nr_populations=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the history object is also always accessible from the abcsmc object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "history is abc.history"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `pyabc.storage.History>` object can, for example,\n",
    "be queried for the posterior probabilities in the populations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "history.total_nr_simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evaluate the model probabililties\n",
    "model_probabilities = history.get_model_probabilities()\n",
    "model_probabilities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now, let's visualize the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_probabilities.plot.bar();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So model 1 is the more probable one. Which is expected as it was centered at 1 and the observed data was also 1, whereas model 0 was centered at 0, which is farther away from the observed data. "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
